tango/integrations/beaker/executor.py

```diff
@@ -49,6 +53,120 @@ def __init__(self, msg: str, experiment_url: str):
         self.experiment_url = experiment_url
 
 
+class ResourceAssignment(NamedTuple):
+    """
+    Resources assigned to a step.
+    """
+
+    cluster: str
+    """
+    The cluster to use to execute the step.
+    """
+
+    resources: TaskResources
+    """
+    The compute resources on the cluster to allocate for execution of the step.
+    """
+
+    priority: Union[str, Priority]
+    """
+    The priority to execute the step with.
+    """
+
+
+class BeakerScheduler(Registrable):
+    """
+    A :class:`BeakerScheduler` is responsible for determining which resources and priority to
+    assign to the execution of a step.
+    """
+
+    default_implementation = "simple"
+    """
+    The default implementation is :class:`SimpleBeakerScheduler`.
+    """
+
+    def __init__(self):
+        self._beaker: Optional[Beaker] = None
+
+    @property
+    def beaker(self) -> Beaker:
+        if self._beaker is None:
+            raise ValueError("'beaker' client has not be assigned to scheduler yet!")
+        return self._beaker
+
+    @beaker.setter
+    def beaker(self, beaker: Beaker) -> None:
+        self._beaker = beaker
+
+    @abstractmethod
+    def schedule(self, step: Step) -> ResourceAssignment:
+        """
+        Determine the :class:`ResourceAssignment` for a step.
+        """
+        raise NotImplementedError()
+
+
+@BeakerScheduler.register("simple")
+class SimpleBeakerScheduler(BeakerScheduler):
+    """
+    The :class:`SimpleBeakerScheduler` just searches the given clusters for one
+    with enough resources to match what's specified by the step's required resources.
+    """
+
+    def __init__(self, clusters: List[str], priority: Union[str, Priority]):
+        super().__init__()
+        self.clusters = clusters
+        self.priority = priority
+        self._latest_clusters_used: List[str] = []
+        if not self.clusters:
+            raise ConfigurationError("At least one cluster is required in 'clusters'")
+
+    def schedule(self, step: Step) -> ResourceAssignment:
+        step_resources = step.resources
+        task_resources = TaskResources(
+            cpu_count=step_resources.cpu_count,
+            gpu_count=step_resources.gpu_count,
+            memory=step_resources.memory,
+            shared_memory=step_resources.shared_memory,
+        )
+        cluster_to_use = self._ensure_cluster(task_resources)
+        if cluster_to_use is None:
+            cluster_to_use = self.clusters[0]
+        # Move cluster to the end of `self._latest_clusters_used`
+        try:
+            self._latest_clusters_used.remove(cluster_to_use)  # type: ignore
+        except ValueError:
+            pass
+        self._latest_clusters_used.append(cluster_to_use)  # type: ignore
+        return ResourceAssignment(
+            cluster=cluster_to_use, resources=task_resources, priority=self.priority
+        )
```

Comment:
Is this method allowed to return a value that says "We can't schedule this one right now."? I'm thinking of a scheduler that doesn't queue more than a few jobs ahead of time, and then waits to see which cluster frees up first. Though once again that's a problem that goes away when Beaker fixes https://github.com/allenai/beaker/issues/2544.

Reply:
At the moment, no. But that would be nice. It would take some refactoring though so should probably be a separate PR.

Reply:
Oh! Just kidding, that was easy: 76e5e4d

Reply:
And cb438bd